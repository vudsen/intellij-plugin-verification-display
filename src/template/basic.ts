import {parseVerifyResult} from "../util";

export type VerifyResult = {
  version: string
  compatibilityProblems?: string
  verificationVerdict?: string
  dependencies?: string
  experimentalApiUsages?: string
  telemetry?: string
  deprecatedUsages?: string
}


const CLASS_PATTERN = /[\w$]+(\.[\w$]+)+((\([\w$ .,]+\))? : [\w$]+(\.[\w$]+)*)?/g
function quoteJavaClass(text: string): string[] {
  const builder: string[] = []
  let result: RegExpExecArray | null = null
  let head = 0
  while ((result = CLASS_PATTERN.exec(text))) {
    const pos = result.index
    builder.push(text.substring(head, pos))
    builder.push(`\`${result[0]}\``)
    head = pos + result[0].length
  }
  if (head < text.length) {
    builder.push(text.substring(head))
  }
  return builder
}

function splitLines(text: string[]): string[] {
  const builder: string[] = []
  if (text.length === 0) {
    return []
  }
  builder.push('- 🐞 ')
  for (let i = 0; i < text.length; i++) {
    const string = text[i]
    const pos = string.indexOf('\n')
    if (pos < 0) {
      builder.push(string)
      continue
    }
    builder.push(string.slice(0, pos))
    if (pos < string.length - 1 || i < text.length - 1) {
      builder.push('\n- 🐞 ')
      builder.push(string.slice(pos + 1))
    }
  }
  return builder
}



/**
 * 将验证结果解析为 markdown
 * @param unzippedRoot 验证输出的文件
 */
export default function parseAsMarkdown(unzippedRoot: string): string {
  const result = parseVerifyResult(unzippedRoot)
  const builder: string[] = ['# Plugin Verifier Check Result']

  for (let verifyResult of result) {
    builder.push(`\n\n## ${verifyResult.version}\n\n**Summary:** ${verifyResult.verificationVerdict}`)
    builder.push('\n\n### Compatibility Problems\n\n')
    if (verifyResult.compatibilityProblems) {
      builder.push(...splitLines(quoteJavaClass(verifyResult.compatibilityProblems)))
    } else {
      builder.push('✅ No compatibility problems found.')
    }
  }
  return builder.join('')
}