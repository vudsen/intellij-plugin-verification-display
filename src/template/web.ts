import {parseVerifyResult} from "../util";

function highlightCompatible(text?: string): string {
  if (!text) {
    return '-'
  }
  const pos = text.indexOf('Compatible')
  if (pos === -1) {
    return '‚ùå ' + text
  }
  return '‚úÖ ' + text
}

export function parseWithWebLink(unzippedRoot: string, visitPath: string): string {
  const result = parseVerifyResult(unzippedRoot)
  const builder: string[] = ["# üìñ Verification Result\n\n| Version | Summary | Links |\n| ------- | ------- | ------ |\n"]
  for (let verifyResult of result) {
    builder.push('| ')
    builder.push(verifyResult.version)
    builder.push(' | ')
    builder.push(highlightCompatible(verifyResult.verificationVerdict?.trim()))
    builder.push(' | ')
    builder.push(`[Detail](${visitPath}/${verifyResult.version}/report)`)
    builder.push(' |\n')
  }
  return builder.join('')
}