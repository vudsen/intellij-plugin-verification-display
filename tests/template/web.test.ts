import {test, expect} from "@jest/globals";
import {parseWithWebLink} from "../../src/template/web";


test('Test basic usage', () => {
  expect(parseWithWebLink('./tests/__resources__/verifyResult', 'https://foo.bar')).toMatchSnapshot()
})