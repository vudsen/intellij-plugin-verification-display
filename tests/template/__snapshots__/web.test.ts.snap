// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test basic usage 1`] = `
"# üìñ Verification Result

| Version | Summary | Links |
| ------- | ------- | ------ |
| IC-232.10335.12 | ‚ùå 4 compatibility problems. 6 usages of experimental API | [Detail](https://foo.bar/IC-232.10335.12/report) |
| IC-233.15619.7 | ‚ùå 1 compatibility problem. 10 usages of experimental API | [Detail](https://foo.bar/IC-233.15619.7/report) |
| IC-241.19416.15 | ‚úÖ Compatible. 2 usages of deprecated API. 10 usages of experimental API | [Detail](https://foo.bar/IC-241.19416.15/report) |
| IC-242.24807.4 | ‚úÖ Compatible. 2 usages of deprecated API. 10 usages of experimental API | [Detail](https://foo.bar/IC-242.24807.4/report) |
| IC-243.25659.39 | ‚úÖ Compatible. 2 usages of deprecated API. 4 usages of experimental API | [Detail](https://foo.bar/IC-243.25659.39/report) |
"
`;
